subprojects/core/src/main/java/org/gradle/internal/classpath/LambdaSerializationTransformer.java

```diff
@@ -0,0 +1,287 @@
+        super.visit(version, access, name, signature, superName, interfaces);
+        this.className = name;
+        this.isInterface = (access & ACC_INTERFACE) != 0;
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+        if (name.equals(DESERIALIZE_LAMBDA) && descriptor.equals(RETURN_OBJECT_FROM_SERIALIZED_LAMBDA)) {
+            hasDeserializeLambda = true;
+            return super.visitMethod(access, RENAMED_DESERIALIZE_LAMBDA, descriptor, signature, exceptions);
+        }
+        return new LambdaFactoryCallRewriter(super.visitMethod(access, name, descriptor, signature, exceptions));
+    }
+
+    @Override
+    public void visitEnd() {
+        if (!lambdaFactories.isEmpty() || hasDeserializeLambda) {
+            generateLambdaDeserializeMethod();
+        }
+        super.visitEnd();
+    }
+
+    private void generateLambdaDeserializeMethod() {
+        // There may be so many lambdas, the resulting deserialization method wouldn't fit into the JVM size limits.
+        // If this happens, we split the generated method into a chain of multiple.
+        // The first split is always $deserializeLambda$. It processes as many lambdas as possible,
+        // and if there are too many, it calls another split method $deserializeLambda0$ as its last operation.
+        // $deserializeLambda0$ may call $deserializeLambda1$ if needed, and so on.
+        PeekingIterator<LambdaFactoryDetails> factoriesIterator = Iterators.peekingIterator(lambdaFactories.iterator());
+        int nextSplitMethodIndex = 0;
+        String currentSplitMethodName = DESERIALIZE_LAMBDA;
+        do {
+            String nextSplitMethodName = String.format("$deserializeLambda%d$", nextSplitMethodIndex);
+            generateSplitLambdaDeserializeMethod(currentSplitMethodName, nextSplitMethodName, factoriesIterator);
+            currentSplitMethodName = nextSplitMethodName;
+            ++nextSplitMethodIndex;
+        } while (factoriesIterator.hasNext());
+    }
+
+    private void generateSplitLambdaDeserializeMethod(String methodName, String nextSplitMethodName, PeekingIterator<LambdaFactoryDetails> factoriesIterator) {
+        CodeSizeEvaluator sizeEvaluator = new CodeSizeEvaluator(visitStaticPrivateMethod(methodName, RETURN_OBJECT_FROM_SERIALIZED_LAMBDA));
+        new MethodVisitorScope(sizeEvaluator) {{
+            Label next = null;
+            while (factoriesIterator.hasNext()) {
+                LambdaFactoryDetails factory = factoriesIterator.peek();
+                Type[] argumentTypes = Type.getArgumentTypes(factory.descriptor);
+
+                int codeSizeSoFar = sizeEvaluator.getMaxSize();
+                if (codeSizeSoFar + getEstimatedSingleLambdaHandlingCodeLength(argumentTypes) + getEstimatedEpilogueLength() >= MAX_CODE_SIZE) {
```

Comment:
An alternative approach that I imagine would be to extract the calls emitting the instructions below into a method receiving the visitor as a parameter, and then execute it first on a `CodeSizeEvaluator` that is not backed by an actual visitor chain leading to the class writer, then check the sizes, and then maybe repeat with the real visitor.

While it's likely slower at runtime, it could probably reduce the amount of "manual" estimation, which would simplify modifying the code. I don't think we are likely to modify it often, though.

Reply:
I though about this, and doing a dry-run is not that much extra work, but it is still work. OTOH, hand-counting has to only be done once, and there's a test to verify its correctness. We can always revisit this decision later, if updating the estimation by hand becomes tedious.

Reply:
Interesting idea. 

> We can always revisit this decision later, if updating the estimation by hand becomes tedious.

👍